lighthouse-core/audits/csp-xss.js

```diff
@@ -0,0 +1,193 @@
+  additionalWarning: 'Additional CSP suggestions are available.',
+  /** Message shown when one or more CSPs are defined in a <meta> tag. Shown in a table with a list of other CSP vulnerabilities and suggestions. "CSP" stands for "Content Security Policy". "CSP" and "HTTP" do not need to be translated. */
+  metaTagMessage: 'The page contains a CSP defined in a <meta> tag. ' +
+    'It is not recommended to use a CSP this way, ' +
+    'consider defining the CSP in an HTTP header.',
+  /** Message shown when a CSP has no syntax errors. Shown in a table with a list of other CSP vulnerabilities and suggestions. "CSP" stands for "Content Security Policy". */
+  noSyntaxErrors: 'No syntax errors.',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+class CspXss extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'csp-xss',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['devtoolsLogs', 'MetaElements', 'URL'],
+    };
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<{cspHeaders: Array<string>, cspMetaTags: Array<string>}>}
+   */
+  static async getRawCsps(artifacts, context) {
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+    const mainResource = await MainResource.request({devtoolsLog, URL: artifacts.URL}, context);
+
+    const cspMetaTags = artifacts.MetaElements
+      .filter(m => {
+        return m.httpEquiv && m.httpEquiv.toLowerCase() === 'content-security-policy';
+      })
+      .flatMap(m => (m.content || '').split(','))
+      .filter(rawCsp => rawCsp.replace(/\s/g, ''));
+    const cspHeaders = mainResource.responseHeaders
+      .filter(h => {
+        return h.name.toLowerCase() === 'content-security-policy';
+      })
+      .flatMap(h => h.value.split(','))
+      .filter(rawCsp => rawCsp.replace(/\s/g, ''));
+
+    return {cspHeaders, cspMetaTags};
+  }
+
+  /**
+   * @param {Finding} finding
+   * @return {LH.Audit.Details.TableItem}
+   */
+  static findingToTableItem(finding) {
+    return {
+      directive: finding.directive,
+      description: getTranslatedDescription(finding),
+    };
+  }
+
+  /**
+   * @param {Array<string>} rawCsps
+   * @return {LH.Audit.Details.TableItem[]}
+   */
+  static collectSyntaxResults(rawCsps) {
+    /** @type {LH.Audit.Details.TableItem[]} */
+    const results = [];
+
+    const syntaxFindingsByCsp = evaluateRawCspForSyntax(rawCsps);
+    for (let i = 0; i < rawCsps.length; ++i) {
+      const items = syntaxFindingsByCsp[i].map(this.findingToTableItem);
+      if (!items.length) {
+        items.push({description: str_(UIStrings.noSyntaxErrors)});
```

Comment:
During analysis of partner sites I noticed that none of the pages have syntax errors (as you'd expect of production sites), but it makes me wonder if we should drop the default "No Syntax Errors" when nothing is wrong.

Reply:
Honestly, that might be a reason to keep it. I originally added it because I thought the CSP sitting on top of a list of vulnerabilities looked out of place. Adding the "No syntax errors" message makes it's purpose obvious. Assuming >99% of CSPs are syntax error free, this confusion is even more concerning to me.

That opinion is pretty subjective though, so I'm willing to remove it if others feel differently.